[REEF-1251] IMRU Driver handlers for fault tolerant

jwang98052 opened the pull request:
* IMRU Driver handler re-write to support fault tolerant
* ServiceAndContextConfigurationProvider refactor and clean up for the updated IMRU driver
* IMRUClient update to use the updated IMRUDriver
* Minor changes in ActiveContextManager
* Test case will follow after REEF-1408 is merged.

JIRA: [REEF-1251](https://issues.apache.org/jira/browse/REEF-1251)
This closes

afchung commented:
I'll take a look.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Client/REEFIMRUClient.cs:
```diff
@@ -104,7 +104,9 @@ internal sealed class REEFIMRUClient : IIMRUClient
-                    .Set(DriverConfiguration.CustomTraceLevel, TraceLevel.Info.ToString())
+                    .Set(DriverConfiguration.OnTaskRunning,
+                        GenericType<IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>>.Class)
+                    .Set(DriverConfiguration.CustomTraceLevel, TraceLevel.Verbose.ToString())
```
This should only be used for debugging.

jwang98052 replied:
yes. I can change it back.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -49,44 +49,74 @@
-        private readonly int _coresForUpdateTask;
-        private readonly int _memoryPerMapper;
-        private readonly int _memoryForUpdateTask;
+        private ConcurrentStack<IConfiguration> _perMapperConfiguration;
```
`_perMapperConfigurationStack`?

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
             {
-                Logger.Log(Level.Info, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                if (!_isFirstTry)
```
`if-else`.

jwang98052 replied:
rest of the code is for both first try and recovery.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+
+                if (_isFirstTry)
+                {
+                    _systemState = new SystemStateMachine();
```
This should be initialized in the constructor.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+
+                if (mappersToRequest == 0)
+                {
+                    Logger.Log(Level.Info, "There is no failed Evaluator in this recovery.");
```
are no failed Evaluators.

afchung replied:
This can be first try right?

jwang98052 replied:
No for first time. I  added the condition.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+                    return;
+                }
+
+                if (_isFirstTry)
```
Why not do everything in one scope?

jwang98052 replied:
There are sequence requirement, 1. got some values such as mappersToRequest, 2. ResetFailedEvaluators. Once reset, we cannot get those values and reset much happen before recovery. 3. Then recovery.

dkm2110 replied:
I guess readability wise what @afchung  says looks more plausible - just write one if(_firsTime) loop and do everything there than in parts like this

jwang98052 replied:
Previously we need to recover both master and mapper evaluators. So first time and later recovery have something in common to share in terns of requesting master and mapper Evaluators. 
Since we are not going to recover master evaluator now, I would just move the code for first time inside IDriverStarted handler, leaving recovery in this method, that would make the code easy to read.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+                    _evaluatorManager.RequestUpdateEvaluator();
+                }
+
+                if (mappersToRequest > 0)
```
Please use `if-else` with `if (mappersToRequest == 0)`

jwang98052 replied:
if (mappersToRequest == 0), donothing?

dkm2110 replied:
I guess he is indicating that throw an exception in that case?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -12,41 +12,43 @@
 // "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 // KIND, either express or implied.  See the License for the
-// specific language governing permissions and limitations
+// specific language governing permissions and limitations_submittedEvaluators
```
Remove.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -12,41 +12,43 @@
     /// </summary>
     /// <typeparam name="TMapInput"></typeparam>
-    /// <typeparam name="TMapOutput"></typeparam>
+    /// <typeparam name="TMapOutput"></typeparam>_submittedEvaluators
```
Remove.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/MaxRetryNumberInRecovery.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Parameters
+{
+    [NamedParameter("Maximum retry number in fault tolerant recovery.", "maxRetryInRecovery", "2")]
```
I think we should make the value 0.

afchung replied:
Add XML doc.

jwang98052 replied:
The initial value start from 1. So I can change the default to 1 meaning no retry be default.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-            var configs =
-                _serviceAndContextConfigurationProvider.GetContextConfigurationForEvaluatorById(allocatedEvaluator.Id);
-            allocatedEvaluator.SubmitContextAndService(configs.Context, configs.Service);
+            Logger.Log(Level.Verbose, string.Format(CultureInfo.InvariantCulture, "AllocatedEvaluator EvaluatorBatchId [{0}], memory [{1}], systemState {2}.", allocatedEvaluator.EvaluatorBatchId, allocatedEvaluator.GetEvaluatorDescriptor().Memory, _systemState.CurrentState));
```
No need for `string.Format`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        SubmitContextAndService(allocatedEvaluator);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Verbose, "Receiving IAllocatedEvaluator event, but system is in FAIL state, ignore it.");
```
Shouldn't you close it?

jwang98052 replied:
No. This is because the code that sets system Fail has executed FailedAction. It has shut down all the allocated evaluators/contexts. If a new IAllocatedEvaluator comes after it, we should not submit anything and no need to close either as the context is not submitted yet. See the comments of this method.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        public void OnNext(IActiveContext activeContext)
         {
-            lock (_completedTasks)
+            Logger.Log(Level.Verbose, string.Format(CultureInfo.InvariantCulture, "Received Active Context {0}, systemState {1}.", activeContext.Id, _systemState.CurrentState));
```
No need for `string.Format`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        public void OnNext(int value)
         {
-            if (AreIMRUTasksCompleted())
+            Logger.Log(Level.Info, string.Format("Received event from ActiveContextManager with NumberOfActiveContexts:" + value));
```
No need for `string.Format`.

afchung replied:
Same for all other log lines.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
         /// </summary>
         /// <param name="value"></param>
-        public void OnNext(IFailedEvaluator value)
+        public void OnNext(int value)
```
Why does this need to be `int`? It doesn't do anything with this value.

jwang98052 replied:
As an IObserver<T>, we need to put something for T. And we need to log it in debug if needed. Do you have a way to not use any T?

dkm2110 replied:
Instead of using IObserver can't u simply use an Action ?

jwang98052 replied:
Conceptually it is an observer. The method is called from observable.

andrey-me replied:
Should it be something like OnNext( IAllContextsReceived value) ?
OnNext(int) doesn't look like readable signature.

jwang98052 replied:
We can only start to submit tasks after all the the active context are received. This is required by IMRU/GC, it also makes the fault tolerant system state transition clean. Please refer to the design document for details which is attached to REEF-1223. 
We had several discussions about this and have decided to use notification. We don't want to keep checking it in IActiveContext  handler. Instead, we want ContextManager to notify Driver to start submitting tasks once all contexts are received. We can achieve the same purpose by many different ways in coding, but this is what we talked and agreed on. Unless you have strong reason to prefer one way to another, then we can bring for discussion again. 

I changed code to use OnNext( IAllContextsReceived value) as you suggested. It added two additional classes, makes a simple issue little bit more complex, but makes the code easy to read.

markusweimer replied:
Shouldn't we have an `Observer<ActiveContext>` which is in turn an `Observable<IEnumerable<ActiveContext>>` to create this state change?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
             }
-            else
+
+            foreach (var mapping in taskIdAndContextMapping)
```
Why do we need two loops here?

jwang98052 replied:
All the tasks must be added to the communication group before calling GetGroupCommTaskConfiguration for any task.

afchung replied:
But you're looping through the same key anyway. Why not do everything in the same loop?

jwang98052 replied:
yes, same key. 
In the second loop, we call GetGroupCommTaskConfiguration(), that requires all the tasks are added to the group first.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        break;
+                    case SystemState.ShuttingDown:
+                    case SystemState.Fail:
+                        _taskManager.RecordRunningTaskDuringSystemFailure(runningTask, TaskManager.CloseTaskByDriver);
```
Does this close the running task?

jwang98052 replied:
Yes. It calls runningTask.Dispose()

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        {
+                            _systemState.MoveNext(SystemStateEvent.AllTasksAreCompleted);
+                            Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "All tasks are completed, systemState {0}", _systemState.CurrentState));
+                            DoneAction();
```
Why do we need this method?

jwang98052 replied:
This method is to dispose all the contexts. We would like to get all the task completed and then dispose contexts together.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+
+        private void FailAction()
+        {
+            ShutDownAllEvaluators();
```
What's the difference between `FailAction()` and `DoneAction()`?

jwang98052 replied:
So far, both dispose all the contexts. DoneAction is for successful scenario and FailAction is for failed and no recovery scenario.

afchung replied:
Please document both methods and explain why they are separate, including what future differences between the two methods may be. If there will be no future differences, please coalesce the methods.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,38 +588,143 @@ public void OnCompleted()
+        /// Checks if the system is recoverable.
+        /// </summary>
+        /// <returns></returns>
+        private bool Recoverable()
```
`IsRecoverable()`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        break;
+
+                    case SystemState.ShuttingDown:
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
```
Should this be called for all cases?

jwang98052 replied:
Ok, I added Fail case. In that case, it fails and we just log and do noting, 
For TaskCompleted state, it has been covered at the begging of the method. 
No more others.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        /// <summary>
+        /// IFailedContext handler. It specifies what to do if Failed Context is received.
+        /// If we get all completed tasks then ignore the failure otherwise throw exception
+        /// Fault tolerant would be similar to FailedEvaluator. It is in TODO
```
What is in TODO?

jwang98052 replied:
IFailedContext is not covered in the design of fault tolerant at this stage.

afchung replied:
`TODO` is a keyword. You should provide a JIRA for it if you want to use `TODO`.

jwang98052 replied:
We haven't defined what to do for this scenario. I would modify the comments for now.

afchung commented:
Done with a first pass.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -57,120 +59,35 @@ internal ServiceAndContextConfigurationProvider(IPartitionedInputDataSet dataset
         /// <param name="evaluatorId"></param>
         /// <returns>Whether failed evaluator is master or not</returns>
-        internal bool RecordEvaluatorFailureById(string evaluatorId)
+        internal bool RemoveEvaluatorIdFromPartitionIdProvider(string evaluatorId)
```
This always returns false. You should make it void?

jwang98052 replied:
True. Changed.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,42 +592,139 @@ public void OnCompleted()
+                var mappersToRequest = _isFirstTry ? _totalMappers : _evaluatorManager.NumberofFailedMappers();
+                _evaluatorManager.ResetFailedEvaluators();
+
+                if (!_isFirstTry && mappersToRequest == 0)
```
if we went to recovery and there is no mapper to request how can this happen?

jwang98052 replied:
If a task fails for some reason, not Evaluator failure, we will  close rest of the tasks and then reach here.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,461 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-        /// Specifies what to do if Failed Context is received.
-        /// An exception is thrown if tasks are not completed.
+        /// IRunningTask handler. The method is called when a task is running. The following action will be taken based on the system state:
+        /// Case WaitingForEvaluator
```
The case description here does not match implementation

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,461 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                    case SystemState.WaitingForEvaluator:
+                        if (IsRecoverable())
+                        {
+                            if (isMaster)
```
I thought we are not recovering from master evaluator failure? IsRecoverable() returns false if master fails. So why this check?

dkm2110 replied:
Never mind this is that case where error happened between IAllocated to IActivecontext

dkm2110 replied:
But you should not use IsRecoverable() here right? It alss includes !IsMaster condition.

jwang98052 replied:
You are right. we should only check if ReachedMaximumNumberOfEvaluatorFailures. It was the code before we decided not to recover master.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,461 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                            _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                        }
+
+                        TryRecovery();
```
This whole function is in lock and once you call StartAction() in TryRecovery() and request for more map evaluators, mapperCount there will always be one right? this is because no other IFaileEvaluator can be received in between because of lock.

dkm2110 replied:
I guess _taskManager.AllInFinalState() in TryRecovery(), makes sure we call StartAction() after all failures etc. have been recorded?

jwang98052 replied:
In recovery, there are two cases, one is no evaluator fail but task failures only, we will only resubmit tasks. The other is some evlautors fail, we will request mappers based on the filed number. This process is locked, no FaieldEvaluaotr event can be received in between. Once we are out of this lock, we do allow to receive failed evaluator again,

jwang98052 replied:
_taskManager.AllInFinalState() in TryRecovery() is to make sure all the tasks are either failed, closed or completed. So that we can resubmit tasks on the existing contexts.

dkm2110 commented:
@jwang98052 I have taken the first pass. One question regarding tests: I know that you are developing test framework for IMRU FT. So is the plan that the driver would be merged first followed by the tests and all. Or you plan to augment this pull request with the tests also. Committing without tests would be very dangerous. :) 

@markusweimer  this pull request is important. Probably you should also take one look?

dkm2110 commented:
@jwang98052  also I just observed above that appveyor build failed for this PR.

markusweimer commented:
I'll have a look in the next iteration. @jwang98052 please ping me such that I do :)

afchung commented:
@jwang98052 why was the PR rebased?

jwang98052 commented:
@afchung I was to merge with REEF-1408 which contains IMRU testing and some bug fixes that would impact the IMRUdriver testing. The testing is needed when I make any changes during the review, I want to make sure the code still pass the tests. As you were the only one who made the review at that time, and this PR will have a long review chain in future I believe, so I think the earlier to rebase the better. Sorry for the inconvenience. 

In fact,  I also want to merge with REEF-1423 once it is in as it fixed a bug that also impact IMRUDriver fault tolerant behavior.

jwang98052 commented:
@dkm2110 REEF-1408 contains tests that can pass for both old and new IMRU drivers. It ensures no regression is introduced. It also contains close event testing for IMRU Tasks which has been updated. REEF-1451 is opened to track the scenario testing where we will add more tests for failed cases.

jwang98052 commented:
@afchung @dkm2110 I have addressed your review comments. Please have another look. Thanks!

afchung commented:
@jwang98052 There are 3 test failures in Appveyor, can you please take a look?

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        public void OnNext(IActiveContext activeContext)
         {
-            lock (_completedTasks)
+            Logger.Log(Level.Verbose, "Received Active Context {0}, systemState {1}.", activeContext.Id, _systemState.CurrentState);
```
Is SystemStateMachine thread safe?

jwang98052 replied:
No

andrey-me replied:
Should Log() be inside a lock then?

jwang98052 replied:
yeah

andrey-me replied:
please check other methods as well, logger statement is often outside of lock

---
In reply to: [67060660](https://github.com/apache/reef/pull/1033#discussion_r67060660) [](ancestors = 67060660)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -49,44 +49,74 @@
+        /// <summary>
+        /// The lock for the driver. 
+        /// </summary>
+        private readonly object _lock = new object();
```
Did we consider not having global lock?
my initial impression is that Driver doesn't store much data, it is just event bus b/w managers and state machine.
Contextmanager is already implementing lock, so if we make state machine and other managers thread safe, this gobal lock may not be needed?

jwang98052 replied:
ActiveContextManager, EvaluatorManager and TaskManager all marked as   [NotThreadSafe]. This was the decision we made during the code review for those PRs. We would like to have a big lock in driver to avoid nested locks, cross use, and potential dead locks.

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+            {
+                if (_taskManager.AreAllTasksCompleted())
+                {
+                    Logger.Log(Level.Verbose, "Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", failedEvaluator.Id);
```
all tasks

jwang98052 replied:
done

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        if (!_evaluatorManager.ReachedMaximumNumberOfEvaluatorFailures())
```
just suggestion: this logic can be hidden in evaluator manager. (would simplify driver code a bit)

jwang98052 replied:
Do you mean the above line? ReachedMaximumNumberOfEvaluatorFailures() is already a method in evaluator manager.

andrey-me replied:
yes, if the manager "knows" maximum number of retries it can decide when to trigger nonRecoverable error.  

---
In reply to: [67057535](https://github.com/apache/reef/pull/1033#discussion_r67057535) [](ancestors = 67057535)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,462 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        break;
+
+                    case SystemState.Fail:
+                        Logger.Log(Level.Error, "Evaluator with Id: {0} failed System is in Fail state.", failedEvaluator.Id);
```
not sure why this logging error message, while other logs are informational.
it might make sense to have this log globally (at the beginning of the handler) to create error record with current system state.

jwang98052 replied:
Done

jwang98052 commented:
@afchung I have addressed your review comments. I also added one scenario where failed evaluator doesn't have associated task id. This can happen if we submit a task, but the evaluator has just failed. Please have another look.

afchung commented:
I'll take another look.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/Org.Apache.REEF.IMRU.csproj:
```diff
@@ -68,12 +68,14 @@ under the License.
     <Compile Include="OnREEF\Driver\DataLoadingContext.cs" />
     <Compile Include="OnREEF\Driver\EvaluatorManager.cs" />
     <Compile Include="OnREEF\Driver\EvaluatorSpecification.cs" />
+    <Compile Include="OnREEF\Driver\IAllContextsReceived .cs" />
```
Fix the file name.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IAllContextsReceived .cs:
```diff
@@ -0,0 +1,27 @@
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    [Private]
+    internal interface IAllContextsReceived
```
No need to mark `[Private]` on an `internal` class.

afchung replied:
Why do we need this class?

jwang98052 replied:
@afchung You asked in your previous comments.You said instead of using IObserver[int], use IObserver[IAllContextsReceived]. I was actually very hesitate as it introduce two more classes.

afchung replied:
That was not me, that was @andrey-me

jwang98052 replied:
Ok. I don't like this class as well, don't want to make simple issue complex basically. I will remove the interface and the  class.

andrey-me replied:
It'll work either way, but OnNext(int) reads like a counter.... have to read through comments and code to understand what the intent is.
I think typed event makes sense here.

---
In reply to: [67778447](https://github.com/apache/reef/pull/1033#discussion_r67778447) [](ancestors = 67778447)

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ActiveContextReceived.cs:
```diff
@@ -0,0 +1,37 @@
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    [Private]
+    internal sealed class ActiveContextReceived : IAllContextsReceived
```
Why do we need this class?

jwang98052 replied:
same as above.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -49,44 +49,74 @@
+        /// <summary>
+        /// It records the number of retry for the recoveries. 
+        /// </summary>
+        private int _numberOfRetryForFaultTolerant = 1;
```
Retries

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+            _maxRetryNumberForFaultTolerant = maxRetryNumberInRecovery;
+
+            _contextManager = new ActiveContextManager(_totalMappers + 1);
+            _contextManager.Subscribe(this);
```
This is dangerous. You're subscribing to an `IObservable` before the constructor is completed.

jwang98052 replied:
A typical way is to pass it in the constructor of ActiveContextManager, result is the same. ActiveContext is an internal class and we know how it is used. In fact it is solely used in IMRUDriver.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-            _perMapperConfiguration = ConstructPerMapperConfigStack(_totalMappers);
+            var updateSpec = new EvaluatorSpecification(memoryForUpdateTask, coresForUpdateTask);
+            var mapperSpec = new EvaluatorSpecification(memoryPerMapper, coresPerMapper);
+            var allowedFailedEvaluators = (int)(failedEvaluatorsFraction * dataSet.Count);
```
use `_totalMappers` instead of `dataSet.Count`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
             [Parameter(typeof(MemoryPerMapper))] int memoryPerMapper,
             [Parameter(typeof(MemoryForUpdateTask))] int memoryForUpdateTask,
             [Parameter(typeof(AllowedFailedEvaluatorsFraction))] double failedEvaluatorsFraction,
```
Why do we use a fraction instead of an actual number?

jwang98052 replied:
if we have 5000 Evaluators, 10 failed, we may still want to recover. If we request 15 Evaluators, 10 failed, we probably don't want to recover. It is relative. It is inherited from original IMRU driver.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        SubmitContextAndService(allocatedEvaluator);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Info, "Receiving IAllocatedEvaluator event, but system is in FAIL state, ignore it.");
```
Shouldn't you close the Evaluator?

jwang98052 replied:
We haven't submitted context on it yet. It is just logically allocated. When all other Evaluators are disposed, the system will shut down.

markusweimer replied:
Still, closing it now is the right thing to do. The allocation can't be given to any other job on the cluster unless we return it.

jwang98052 replied:
Done.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
-                Logger.Log(Level.Info,
-                    string.Format("Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
-                return;
+                // Change the system state to SubmittingTasks
```
This comment is confusing. Are we changing to `SubmittingTasks` or `AllContextsAreReady`?

jwang98052 replied:
the state transition is currentState with an Event => newState, here AllContextsAreReady is the event. I have modified the comment to make it clear.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        }
+                        else
+                        {
+                            Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
```
`Error` is probably more appropriate here.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -282,42 +590,127 @@ public void OnCompleted()
+        /// </summary>
+        private void TryRecovery()
+        {
+            if (_taskManager.AllInFinalState())
```
`AreAllTasksInFinalState`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ActiveContextManager.cs:
```diff
@@ -71,7 +71,7 @@ internal int NumberOfMissingContexts
         /// <param name="activeContextObserver"></param>
         /// <returns></returns>
-        public IDisposable Subscribe(IObserver<IDictionary<string, IActiveContext>> activeContextObserver)
+        public IDisposable Subscribe(IObserver<IAllContextsReceived> activeContextObserver)
```
Is this an `Observable`? Why do we need this `Subscribe`?

afchung replied:
Can we just inject an object?

jwang98052 replied:
For the few *Managers, EvaluatorManager needs two EvaluatorSpecification from the same class, Tang doesn't support. ActiveContextManager needs totalContexts that is got from IMRUDriver dataSet.Count and IPartitionedInputDataSet is injected in IMRUDriver, unless we inject IPartitionedInputDataSet in the ActiveContextManager constructor, which seems unnecessary. Similar things are for the parameters used in TaskManager. To make it consistent, creating their instances in IMRUDriver is clean.

jwang98052 replied:
ActiveContextManager is an Observable. Theoretically it should implement IObservable<>

afchung commented:
@jwang98052 done with another pass. Sorry I'm a bit lost on the topic since the PR was rebased and it was hard to follow previous discussions.

jwang98052 commented:
@afchung I have addressed your review comments. Can you have another look?

afchung commented:
I'll take another pass.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
         public void OnNext(IDriverStarted value)
         {
-            RequestUpdateEvaluator();
-            //// TODO[REEF-598]: Set a timeout for this request to be satisfied. If it is not within that time, exit the Driver.
```
The TODO still applies, right?

jwang98052 replied:
Yes, added the  comments back.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        Logger.Log(Level.Info, "Receiving IAllocatedEvaluator event, but system is in FAIL state, ignore it.");
+                        break;
+                    default:
+                        UnexpectedState(allocatedEvaluator.Id, "IAllocatedEvaluator");
```
Wouldn't it be easier to just throw an exception here? That way, we fail the Driver and get a full exception stack.

jwang98052 replied:
UnexpectedState trows exception. As the same code is used in multiple places, I just wrap it in a method for consistent error message

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        /// and merges it with the task configuration.
+        /// When all the tasks are added, calls TaskManager to SubmitTasks().
+        /// </summary>
+        private void SubmitTasks()
```
Shouldn't this method move to `TaskManager` or such? We don't seem to get much benefit from the state machine in terms of code complexity here.

markusweimer commented:
I did half a pass. My concern is that the `IMRUDriver` doesn't seem to loose complexity as part of the move to the various `Manager` classes. In fact, it seems to be gaining complexity and state.

I was hoping for a significant reduction in complexity when we do things via Rx operators. For instance, we could have something like this:

```
class ContextManager: Observer<ActiveContext>, Observer<FailedContext>, Observable<IEnumerable<ActiveContext>>
{
}
```
This class receives all events as they pertain to contexts and emits the `IEnumerable<ActiveContext>` whenever it has received enough `ActiveContext` events. Similarly, it could have access to a `IMRUStateManager` which keeps tabs on the overall state of the system. If an event violates the state machine assumption, `ContextManager` would call `OnError` for its subscribers.

Similar approaches can the thought of for `EvaluatorManager`, `TaskManager`, ...

Maybe I am being naive, but this approach should encapsulate the mutable state (and therefore complexity) of IMRUDriver into smaller units, which makes it much easier to test and debug. Right now, I fear we use the various `*Manager` classes as collection objects only, which keeps all of the logic in the Driver.

jwang98052 commented:
@markusweimer I thought about the way you mentioned duting teh design. Basically EvaluatorManager is also the AllocatedEvaluatorHandler and FailedEvaluatorHandler, TaskManager is also RuningTaskHandler and  FailedTaskHandler and responsible for submit tasks, etc. 

The pros are, driver is simple, and the managers can be reused in other fault tolerant if there are abstracted good enough. 
The cons are all those managers need to know each other. For example, in FailedEvaluatorHandler, EvaluatorManager needs to talk to TaskManager for the failed task wrapped in the FailedEvaluator. With the current approach, driver knows every one and control the system state. All the managers are responsible for managing their own entities, they don't talk to each other and they also don't know the system state. 
Currently, Driver managers the recover scenarios. Recover can be triggered by either FailedEvaluator or FailedTask, or even completed task,. If changing the approach, we might need a RecoveryManager, Again, all the managers need to know it and it needs to know all the other managers. 
As the test ability, if all the managers need to talk to each other, it is hard to test individually. Currently since they are independent, I have tests for each of them independently. 

So either driver centralized way, or manager distributed way, each has pros and corns.

jwang98052 commented:
I allocated today to try the refactor, the idea is to make 
EvaluatorManager: IObserver<IDriverStarted>, IObserver<IAllocatedEvaluator>, IObserver<IFailedEvaluator>
ActiveContextManager : IDisposable, IObservable<IEnumerable<IActiveContext>>, IObserver<IActiveContext>, IObserver<IFailedContext>
TaskManager : IObserver<ICompletedTask>, IObserver<IFailedTask>, IObserver<IRunningTask>, IObserver<IEnumerable<IActiveContext>>

That basically means, copy paste all the handlers in IMRUDriver to those three managers. IMRUDriver is almost gone, however, the same code goes to those managers. There is nothing reduced. 

And it introduced a lot of complexity. 
1. All the managers need to know each other. e.g. in FailedEvaluator, it needs to inform Context manger to remove failed Context in the failed evaluator and Task Manager to record failed task. In Task Manager, it needs to ask Evaluator manager if EvaluatorId in IActiveContext is a master Evaluator. 
2. Since those are handlers, they must be injectable. That's fine. The result is most of the parameters in IMRUDriver constructor are copied to all the three managers. make every constructor complex. 
3. As some of the code needs to use <TMapInput, TMapOutput, TResult, TPartitionType>, it ends up each manager has to be generic with those Ts. Make the code looks complex.
4. The code IsRecoverable() and RecoveryAction() etc need to access all the three managers. So it may end up a FaultManager. As the three managers need to reference this class as well. I tried future injection, it doesn't work out. 
I am half way of the refactoring, so far I don't see the code is significantly simplified but more complex. 
Given the time constrains, I would suggest to complete the review of the current version. Make the logic work first. If we still like to refactor it, we can leave it as a separate Jira as it would be purely refactor.

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -49,44 +49,74 @@
+        /// <summary>
+        /// It records the number of retry for the recoveries. 
+        /// </summary>
+        private int _numberOfRetriesForFaultTolerant = 1;
```
>private int _numberOfRetriesForFaultTolerant = 1; [](start = 8, length = 49)

Should number of retries be initialized with 0?

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -98,181 +128,461 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        }
+        #endregion IFailedTask
 
         public void OnError(Exception error)
```
>OnError(Exception error) [](start = 20, length = 24)

Should there be handling on error ?

markusweimer commented:
Is this ready for another round of reviews?

jwang98052 commented:
The code now is branched to REEF-1251-t on tcNickolas/reef. Any way we can replace it with this PR?

markusweimer commented:
Hmm, I believe all of the issues raised here have been addressed. Hence, closing this, rebasing and making a clean new PR would be OK, right?

tcNickolas commented:
Yes, we will close this and open a new PR once we're ready (we're still working on debugging and adding tests). I'll try to rebase that branch in two commits - ones which have been reviewed here and new changes, to simplify review

jwang98052 commented:
The newly rebased change is published in
https://github.com/apache/reef/pull/1087
@markusweimer @dkm2110 Please review it. 
I am closing this one for now.
